lib/pwned_password_downloader.rb

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env ruby
+require 'fileutils'
+require 'net/http/persistent'
+require 'retries'
+require 'ruby-progressbar'
+require 'ruby-progressbar/outputs/null'
+
+class PwnedPasswordDownloader
+  attr_reader :destination,
+              :num_threads,
+              :keep_threshold,
+              :output_progress
+
+  alias_method :output_progress?, :output_progress
+
+  RANGE_API_ROOT = 'https://api.pwnedpasswords.com/range/'
+  SHA1_LENGTH = 40
+  HASH_PREFIX_LENGTH = 5
+  OCCURRENCE_OFFSET = SHA1_LENGTH - HASH_PREFIX_LENGTH + ':'.length
+
+  def initialize(
+    destination: 'tmp/pwned',
+    num_threads: 64,
+    keep_threshold: 30,
+    output_progress: true
+  )
+    @destination = destination
+    @num_threads = num_threads
+    @keep_threshold = keep_threshold
+    @output_progress = output_progress
+  end
+
+  def run!(start: '00000', finish: 'FFFFF')
+    (start.to_i(16)..finish.to_i(16)).each do |prefix_num|
+      queue << prefix_num.to_s(16).upcase.rjust(HASH_PREFIX_LENGTH, '0')
+    end
+
+    FileUtils.mkdir_p(destination)
+
+    progress_bar = ProgressBar.create(
+      title: 'Downloading...',
+      total: queue.size,
+      output: output_progress? ? $stdout : ProgressBar::Outputs::Null,
+      format: '[ %t ] %p%% %B %a (%E)',
+    )
+
+    failed_prefixes = Queue.new
+
+    [num_threads, queue.size].min.times do
+      Thread.new do |thread_id|
+        net_http = Net::HTTP::Persistent.new(name: "thread_id_#{thread_id}")
+
+        while (prefix = queue.pop)
+          if already_downloaded?(prefix)
+            progress_bar.increment
+            next
+          end
+
+          begin
+            write_one(
+              prefix:,
+              content: with_retries(max_tries: 5, rescue: Socket::ResolutionError) do
+                download_one(prefix:, net_http:)
+              end,
+            )
+          rescue
+            failed_prefixes << prefix
+          else
+            progress_bar.increment
+          end
+        end
+      ensure
+        net_http.shutdown
+      end
+    end
+
+    wait_for_progress until progress_bar.finished? || failed_prefixes.present?
```

Comment:
`.present?` gets filled in by activesupport... I'm only seeing [`#empty?`](https://ruby-doc.org/core-3.0.2/Queue.html#method-i-empty-3F) as official API? We aren't loading Activesupport in this script
```suggestion
    wait_for_progress until progress_bar.finished? || !failed_prefixes.empty?
```

Reply:
Good point, I'll update.